bootable-jar/WFLY-13305_Bootable_Jar.adoc

```diff
@@ -0,0 +1,338 @@
+
+==== Arguments handling
+
+Arguments specific to bootable JAR:
+
+|===
+|Option |Description
+|--help
+|Display help then exit
+
+|--deployment=<path to WAR/JAR/EAR file or exploded deployment directory>
+|Application to install in the hollow JAR. Adding a deployment to an bootable JAR already containing a deployment is invalid.
+
+|--install-dir=<path to directory to install server in>
+|By default a new TEMP directory is created. TEMP directory location is controlled by the Java VM (call to ```Files.createTempDirectory```).
+|===
+
+Server arguments:
+
+|===
+|Option |Description
+|-b[interface]=<value>
+|Set system property jboss.bind.address.<interface> to the given value
+
+|-b=<value>
+|Set system property jboss.bind.address to the given value
+
+|-D<name>[=<value>] 
+|Set a system property. The system properties are set by the server. They are not set by the bootable JAR JVM.
+
+|-u=<value>
+|Set system property jboss.default.multicast.address to the given value.
+
+|--version
+|Print version and exit.
+
+|-S<name>[=value]
+|Set a security property
+
+|--properties=<url>
+|Load system properties for the given url
+|===
+
+==== Developer experience
+
+Having to rebuild a server and package a bootable JAR for each code change is not a valid approach. 
+The plugin should offer a development mode allowing to make development using bootable JAR an efficient task. 
+We are defining here a workflow that leverages the hollow JAR packaging and server deployment scanner capabilities.
+
+===== Dev mode
+
+* A dev server is an hollow server scanning the directory target/deployments.
+* A dev app is the primary artifact copied to the target/deployments dir.
+
+The plugin attempts to force the provisioning of the deployment scanner in dev mode:
+
+* If deployment-scanner is excluded, remove it from exclusion. A warning message is displayed advertising that the provisioning of the deployment-scanner is enforced.
+* Add the deployment-scanner layer to the set of layers. An info message is displayed advertising that the deployment-scanner is provisioned.
+* If a provisioning.xml file is set (and no layers override it), a warning is displayed advertising that we can’t enforce the presence of the deployment-scanner.
+* If the deployment-scanner is not in the config, the CLI operation will abort the plugin execution and log an error message.
+
+===== Workflow examples
+
+Description of the 2 workflows (with and without dev mode). These examples could be designed differently according to the content of the pom file.
+
+Nominal mode, full repackaging of the server done for each rebuild, server restarted after each rebuild (not viable):
+
+* mvn package ⇒ full repackaging
+* mvn wildfly-jar:run
+* (kill synchronous execution).
+* User make changes
+* mvn package ⇒ full repackaging
+* ...
+
+Dev mode, server built/started once, app automatically re-deployed.
```

Comment:
When user provide CLI scripts they want to apply, in which phase they will be applied in dev mode?

Reply:
In dev mode, the server is not built, so CLI scripts are not applied.

Reply:
Thanks! IIUC they should be executed when running `mvn wildfly-jar:dev` - is that correct?

Reply:
Yes, when the server is built, CLI scripts are run.